general/mm.cpp

```diff
@@ -0,0 +1,366 @@
+{
+   MFEM_ASSERT(!Known(maps, adrs), "Adrs is an already known address!");
+   for (mm_iterator_t mem = maps->memories->begin();
+        mem != maps->memories->end(); mem++)
+   {
+      const void *b_adrs = mem->first;
+      if (b_adrs > adrs) { continue; }
+      const void *end = (char*)b_adrs + mem->second->bytes;
+      if (adrs < end) { return b_adrs; }
+   }
+   return NULL;
+}
+
+// *****************************************************************************
+static const void* InsertAlias(const mm_t *maps,
+                               const void *base,
+                               const void *adrs)
+{
+   memory_t *mem = maps->memories->at(base);
+   const size_t offset = (char*)adrs - (char*)base;
+   const alias_t *alias = new alias_t(mem, offset);
+   maps->aliases->operator[](adrs) = alias;
+   mem->aliases.push_back(alias);
+   return adrs;
+}
+
+// *****************************************************************************
+// * Tests if adrs is an alias address
+// *****************************************************************************
+static bool Alias(const mm_t *maps, const void *adrs)
+{
+   const auto found = maps->aliases->find(adrs);
+   const bool alias = found != maps->aliases->end();
+   if (alias) { return true; }
+   const void *base = IsAlias(maps, adrs);
+   if (!base) { return false; }
+   InsertAlias(maps, base, adrs);
+   return true;
+}
+
+// *****************************************************************************
+// __attribute__((unused)) // VS doesn't like this in Appveyor
+static void debugMode(void)
+{
+   dbg("\033[1K\r%sMM %sHasBeenEnabled %sEnabled %sDisabled \
+%sCPU %sGPU %sPA %sCUDA %sOCCA",
+       config::usingMM()?"\033[32m":"\033[31m",
+       config::gpuHasBeenEnabled()?"\033[32m":"\033[31m",
+       config::gpuEnabled()?"\033[32m":"\033[31m",
+       config::gpuDisabled()?"\033[32m":"\033[31m",
+       config::usingCpu()?"\033[32m":"\033[31m",
+       config::usingGpu()?"\033[32m":"\033[31m",
+       config::usingPA()?"\033[32m":"\033[31m",
+       config::usingCuda()?"\033[32m":"\033[31m",
+       config::usingOcca()?"\033[32m":"\033[31m");
+}
+
+// *****************************************************************************
+// * Adds an address
+// *****************************************************************************
+void* mm::Insert(void *adrs, const size_t bytes)
+{
+   if (!config::usingMM()) { return adrs; }
+   if (config::gpuDisabled()) { return adrs; }
+   const bool known = Known(maps, adrs);
+   MFEM_ASSERT(!known, "Trying to add already present address!");
+   dbg("\033[33m%p \033[35m(%ldb)", adrs, bytes);
+   memories->operator[](adrs) = new memory_t(adrs,bytes);
+   return adrs;
+}
+
+// *****************************************************************************
+// * Remove the address from the map, as well as all the address' aliases
+// *****************************************************************************
+void *mm::Erase(void *adrs)
```

Comment:
@camierjs Is there a reason that this function returns the same pointer that it was passed?

Reply:
No particular reason, the `free` function could use `ptr`, right.

Reply:
Most likely to mirror `mm::Insert(void*, const size_t)`, which returns a pointer. I changed this in `jdahm/okina-mm-dev`.